### ISSUE TITLE
Failed to fetch L1 block info and receipts endlessly when get null from eth_getBlockReceipts API once

### ISSUE THREAD

**Issue by LilkkJerry** [ethereum-optimism/optimism#9274]
**Describe the bug**
I am now wrapping several nodes into one L1 endpoint to ensure the high availability,  in this situation the request will be handled by any node in this pool. Sometimes i found the op-node will stop working and print this over and over again:
**failed to fetch L1 block info and receipts: invalid receipts: got 0 receipts but expected 68**
I think it's because some of the nodes in the pool have not synchronized the newest block. But, it will not recover even if all of the nodes have already synced the block. After troubleshoot I think it is because the cache mechanism of FetchReceipts:

	func (p *CachingReceiptsProvider) FetchReceipts(ctx context.Context, block eth.BlockID, txHashes []common.Hash) (types.Receipts, error) {
	if r, ok := p.cache.Get(block.Hash); ok {
		return r, nil
	}

	mu := p.getOrCreateFetchingLock(block.Hash)
	mu.Lock()
	defer mu.Unlock()
	// Other routine might have fetched in the meantime
	if r, ok := p.cache.Get(block.Hash); ok {
		// we might have created a new lock above while the old
		// fetching job completed.
		p.deleteFetchingLock(block.Hash)
		return r, nil
	}

	r, err := p.inner.FetchReceipts(ctx, block, txHashes)
	if err != nil {
		return nil, err
	}
	p.cache.Add(block.Hash, r)
	// result now in cache, can delete fetching lock
	p.deleteFetchingLock(block.Hash)
	return r, nil

The response should not be put into cache if the length of receipts is 0, because we don't know if it's real response or it's beacuse the block has not arrived.
Currently I can only fix this by restart the op-node.

**Comment by tobidae-cb** [comment]
Noticed something similar, I'm on op-node 1.4.3. What's interesting is that I have 4 different pods that use the same L1 RPC url and not all of them fail at the exact same l1 block. It's currently a hit or miss

```
optimism-snapshot-0 op-node t=2024-02-01T06:03:20+0000 lvl=warn msg="Derivation process temporary error"     attempts=1 err="temp: failed to fetch receipts of L1 block 0x8c097b788ea6560e3141494d6b1b57060d03ae47f3ab5d89f87dfae447fb9944:19131314 (parent: 0x0635b596a6ab9cb4b6185e7195f4b31a43425625483936186b2e08e5c0f60242:19131313) for L1 sysCfg update: invalid receipts: got 0 receipts but expected 152"
```

Full logs
```shell
optimism-snapshot-0 op-node t=2024-02-01T18:07:03+0000 lvl=info msg="disconnected from peer"                 peer=16Uiu2HAm6h61WPbW7tySjVqJsF1ArvQ5G2HWSGicNTh31TBdHcVB addr=/ip4/116.202.215.20/tcp/9222
optimism-snapshot-0 op-node t=2024-02-01T18:07:03+0000 lvl=info msg="connected to peer"                      peer=16Uiu2HAkvspxvzuCxyChrLF3c3YHZk1JjLJCvXAmryGGT1AqTapu addr=/ip4/176.9.111.210/tcp/9222
optimism-snapshot-0 op-node t=2024-02-01T18:07:03+0000 lvl=info msg="Starting P2P sync client event loop"    peer=16Uiu2HAkvspxvzuCxyChrLF3c3YHZk1JjLJCvXAmryGGT1AqTapu
optimism-snapshot-0 op-node t=2024-02-01T18:07:03+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAmUXhiDnUaKFDUh5pf3JNskdSWhxuKz1h1ua2DA1X96teK
optimism-snapshot-0 op-node t=2024-02-01T18:07:03+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAm5SosPy758VbgU1JfVoM6ytcv1J4KhfUXEibwzf4jeYXA
optimism-snapshot-0 op-node t=2024-02-01T18:07:03+0000 lvl=info msg="Received signed execution payload from p2p" id=0xf5afca046305624f2866afeaf92daf385d4aa7054e475f84130be5b96254b127:115606023 peer=16Uiu2HAm1A4meETRCdaVdCnxtqFA7AnFDtsKvJYmD1gxyJ7VR1dY
optimism-snapshot-0 op-node t=2024-02-01T18:07:03+0000 lvl=info msg="Optimistically queueing unsafe L2 execution payload" id=0xf5afca046305624f2866afeaf92daf385d4aa7054e475f84130be5b96254b127:115606023
optimism-snapshot-0 op-node t=2024-02-01T18:07:04+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAkzJEZcmp2T8mEW1CAWBF5BaJCFQN2vdgyGi3phu6Vsjs9
optimism-snapshot-0 op-node t=2024-02-01T18:07:04+0000 lvl=info msg="disconnected from peer"                 peer=16Uiu2HAm2S7iFb15gjN4KEkYk35pCnQ5HTWfmk1HzXRWwydSeoAJ addr=/ip4/173.231.40.194/tcp/31367
optimism-snapshot-0 op-node t=2024-02-01T18:07:04+0000 lvl=warn msg="Derivation process temporary error"     attempts=2 err="temp: failed to fetch receipts of L1 block 0x8c097b788ea6560e3141494d6b1b57060d03ae47f3ab5d89f87dfae447fb9944:19131314 (parent: 0x0635b596a6ab9cb4b6185e7195f4b31a43425625483936186b2e08e5c0f60242:19131313) for L1 sysCfg update: invalid receipts: got 0 receipts but expected 152"
optimism-snapshot-0 op-node t=2024-02-01T18:07:05+0000 lvl=info msg="Received signed execution payload from p2p" id=0xdefac110a701cd8e14c2577b948cf71a116e0f0f49cba8919eb1914c4ae2bca9:115606024 peer=16Uiu2HAm8HjPrubW1mjGgqSs1doXuTLD5nDzrSAFMkLn6TGRnH9C
optimism-snapshot-0 op-node t=2024-02-01T18:07:05+0000 lvl=info msg="Optimistically queueing unsafe L2 execution payload" id=0xdefac110a701cd8e14c2577b948cf71a116e0f0f49cba8919eb1914c4ae2bca9:115606024
optimism-snapshot-0 op-node t=2024-02-01T18:07:06+0000 lvl=warn msg="failed p2p sync request"                peer=16Uiu2HAkyfu9f78iAbNzq6bohjL4MXWQBkQ31WxDEb5M73WJs78F num=115,606,018 err="failed to open stream: failed to negotiate protocol: protocols not supported: [/opstack/req/payload_by_number/10/0]"
optimism-snapshot-0 op-node t=2024-02-01T18:07:07+0000 lvl=info msg="Received signed execution payload from p2p" id=0x62951ff3535c8cde29b08ab66b17c4c84386559e59dc595e3550eff0046615d3:115606025 peer=16Uiu2HAm8HjPrubW1mjGgqSs1doXuTLD5nDzrSAFMkLn6TGRnH9C
optimism-snapshot-0 op-node t=2024-02-01T18:07:07+0000 lvl=info msg="Optimistically queueing unsafe L2 execution payload" id=0x62951ff3535c8cde29b08ab66b17c4c84386559e59dc595e3550eff0046615d3:115606025
optimism-snapshot-0 op-node t=2024-02-01T18:07:08+0000 lvl=warn msg="Derivation process temporary error"     attempts=3 err="temp: failed to fetch receipts of L1 block 0x8c097b788ea6560e3141494d6b1b57060d03ae47f3ab5d89f87dfae447fb9944:19131314 (parent: 0x0635b596a6ab9cb4b6185e7195f4b31a43425625483936186b2e08e5c0f60242:19131313) for L1 sysCfg update: invalid receipts: got 0 receipts but expected 152"
optimism-snapshot-0 op-node t=2024-02-01T18:07:10+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAkuSMo7R59r39ZoFrVUrokQLrvrNsx4K2h6QnSJ69cu2fk
optimism-snapshot-0 op-node t=2024-02-01T18:07:10+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAmJZhUCVoAYBYS7XygYrq53MLFXFv1pRRXGw9EMj8wqn4m
optimism-snapshot-0 op-node t=2024-02-01T18:07:10+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAmPhZ5EJavkVcwwYsBj2MKLKaN1j53RvMnWo5ZS7kfs9WW
optimism-snapshot-0 op-node t=2024-02-01T18:07:10+0000 lvl=warn msg="failed p2p sync request"                peer=16Uiu2HAm3EFdZW6NY1bCrdgxApQTT4crm6trebU7vb6Hywp5g6hc num=115,606,018 err="failed to open stream: failed to negotiate protocol: protocols not supported: [/opstack/req/payload_by_number/10/0]"
optimism-snapshot-0 op-node t=2024-02-01T18:07:10+0000 lvl=info msg="Received signed execution payload from p2p" id=0x75dcee6c522c311c339923200a243433b967c4a58ee491386d15cf0dfc62dc0a:115606026 peer=16Uiu2HAm1A4meETRCdaVdCnxtqFA7AnFDtsKvJYmD1gxyJ7VR1dY
optimism-snapshot-0 op-node t=2024-02-01T18:07:10+0000 lvl=info msg="Optimistically queueing unsafe L2 execution payload" id=0x75dcee6c522c311c339923200a243433b967c4a58ee491386d15cf0dfc62dc0a:115606026
optimism-snapshot-0 op-node t=2024-02-01T18:07:11+0000 lvl=info msg="disconnected from peer"                 peer=16Uiu2HAkvAmZsULW3K1uij4g1nU3ypRDELaLP1sD2wjuqCsA4TgS addr=/ip4/136.243.16.41/tcp/9222
optimism-snapshot-0 op-node t=2024-02-01T18:07:11+0000 lvl=info msg="Received signed execution payload from p2p" id=0xd5ffcbd47f133d007b4c16abf2a63fac59a13449bc820f2aa994564b8b12a2d0:115606027 peer=16Uiu2HAm1A4meETRCdaVdCnxtqFA7AnFDtsKvJYmD1gxyJ7VR1dY
optimism-snapshot-0 op-node t=2024-02-01T18:07:11+0000 lvl=info msg="Optimistically queueing unsafe L2 execution payload" id=0xd5ffcbd47f133d007b4c16abf2a63fac59a13449bc820f2aa994564b8b12a2d0:115606027
optimism-snapshot-0 op-node t=2024-02-01T18:07:13+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAkzSeapxGvEZvxxj1EE2orxehRz8KL6A7572Jd1rzAFRik
optimism-snapshot-0 op-node t=2024-02-01T18:07:13+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAmFCGhfVNtEUAVYtTgqjjkSnUXWJRio3oG8seV6bGy4pse
optimism-snapshot-0 op-node t=2024-02-01T18:07:13+0000 lvl=info msg="Received signed execution payload from p2p" id=0x3f19d3aa4c4357f194a7a9c9b1ef34ab04de23e1637d4beeed6eb31e957882ed:115606028 peer=16Uiu2HAm8HjPrubW1mjGgqSs1doXuTLD5nDzrSAFMkLn6TGRnH9C
optimism-snapshot-0 op-node t=2024-02-01T18:07:13+0000 lvl=info msg="Optimistically queueing unsafe L2 execution payload" id=0x3f19d3aa4c4357f194a7a9c9b1ef34ab04de23e1637d4beeed6eb31e957882ed:115606028
optimism-snapshot-0 op-node t=2024-02-01T18:07:14+0000 lvl=info msg="attempting connection"                  peer=16Uiu2HAmB4NjKRgNZjyfwXEyTzNUgrzwCouaqAv7XRtE9T4YKiNN
optimism-snapshot-0 op-node t=2024-02-01T18:07:14+0000 lvl=warn msg="failed p2p sync request"                peer=16Uiu2HAm1hmjDZjF27FP1oCQBS8ND9DWXZX1yyRNexTyr7eC5WL3 num=115,606,018 err="failed to open stream: failed to negotiate protocol: protocols not supported: [/opstack/req/payload_by_number/10/0]"
```

**Comment by sebastianst** [comment]
I can confirm that this is a bug in case that the inner receipts fetcher returns a `nil` error together with an empty receipts list. Here, the wrong assumption was made that
```
r, err := p.inner.FetchReceipts(ctx, block, txHashes)
```
would always return a non-`nil` error if the receipts couldn't be retrieved in full. But it is happening, which is a separate issue to investigate. 

We will fix this behavior soon to make sure we only cache healthy receipts.

It would be helpful to know what rpc kinds you are using.

@tobidae-cb I don't see relevant errors in your full log dump? Only the one line log.

**Comment by tobidae-cb** [comment]
> @tobidae-cb I don't see relevant errors in your full log dump? Only the one line log.
I've edited the logs to contain relevant logs.

For the Eth L1 RPC, I use an eth archival deployment that has 3 pods with separate datastores. The routing method round-robin. On a side note, would an eth full deployment work for op-node? It used to be able to work but for the past couple months, I've had to use an archival deployment

Here are some of the op-node details
```
op-geth - v1.101305.0 - c50337a60a1309a0f1dca3bf33ed1bb38c46cdd7
op-node - v1.4.3 - 28c359620f2b28238163eced79b39335d3a40661
op-legacy-geth - efea631cf19d57d3a64374cb81ce545f958bd38f
```

```
# op-geth
/usr/local/bin/geth \
    --ws \
    --ws.port="8546" \
    --ws.addr=0.0.0.0 \
    --ws.origins="*" \
    --http \
    --http.port="8545" \
    --http.addr=0.0.0.0 \
    --http.vhosts="*" \
    --http.corsdomain="*" \
    --http.api=web3,debug,eth,txpool,net,engine \
    --authrpc.addr=0.0.0.0 \
    --authrpc.jwtsecret="/data/secrets/jwt.txt" \
    --authrpc.port="8551" \
    --authrpc.vhosts="*" \
    --datadir=/data/op-geth \
    --verbosity="4" \
    --rollup.disabletxpoolgossip=true \
    --rollup.historicalrpc="http://localhost:8547/" \
    --rollup.sequencerhttp="https://mainnet-sequencer.optimism.io/" \
    --rollup.historicalrpctimeout=240s \
    --nodiscover \
    --syncmode=full \
    --maxpeers=100 \
    --gcmode=archive \
    --snapshot=false \
    --cache="16384" \
    --op-network="op-mainnet"  \
    --rollup.halt=minor

# op-legacy-geth
/usr/local/bin/geth \
    --datadir=/data/op-legacy-geth \
    --config="/app/assets/l2geth.toml" \
    --gcmode=archive \
    --port=29888

# op-node
/usr/local/bin/op-node \
    --l1="<redacted eth archival node>" \
    --l2="ws://localhost:8551/" \
    --l2.jwt-secret="/data/secrets/jwt.txt" \
    --l1.trustrpc="true" \
    --p2p.disable="false" \
    --network="mainnet" \
    --rpc.addr=0.0.0.0 \
    --rpc.port="9545" \
    --p2p.peerstore.path memory \
    --p2p.discovery.path memory \
    --p2p.priv.path "/data/secrets/p2p.txt" \
    --rollup.halt=minor  \
    --rollup.load-protocol-versions=true
```

**Comment by LilkkJerry** [comment]
> I can confirm that this is a bug in case that the inner receipts fetcher returns a `nil` error together with an empty receipts list. Here, the wrong assumption was made that
> 
> ```
> r, err := p.inner.FetchReceipts(ctx, block, txHashes)
> ```
> 
> would always return a non-`nil` error if the receipts couldn't be retrieved in full. But it is happening, which is a separate issue to investigate.
> 
> We will fix this behavior soon to make sure we only cache healthy receipts.
> 
> It would be helpful to know what rpc kinds you are using.
> 
> @tobidae-cb I don't see relevant errors in your full log dump? Only the one line log.

I am a developer from Huawei Cloud. we provide HA capability which allow user to wrap several nodes into 1 endpoint. Welcome to ask me for more details :)

**Comment by tobidae-cb** [comment]
Hi, what's the status on a potential fix here?

**Comment by sebastianst** [comment]
This is fixed as of `op-node/v1.5.1`. Please upgrade to latest `op-node/v1.6.1`.

**Comment by askucher** [comment]
got it in 1.9.1

```
184|op-node  | t=2024-10-06T23:26:55+0000 lvl=warn msg="Engine temporary error" err="temp: failed to fetch receipts of L1 block 0x26e6f7c5c14f2a35e3513d1f9cfe43148577141f9a8d1717cfc56bb95deba733:56 (parent: 0xb19ffb7493ddc69ea00e4f98b110569b376618d59d86361c8b513a61f0c27816:55) for L1 sysCfg update: failed to fetch list of receipts: expected receipt root 0xbe2657e2ef940ad2bb8cf7745343bc209ffe16cdb7fc3c369973e49a8b576e64 but computed 0x6e4bec7d9ad661e3df512985bd021125139ea67a60cd2aef36449cc7775ecab6 from retrieved receipts" 
```
